WordPress/gutenberg
# Block API: Add support for icons for block categories

## Description
This PR adds support for icons for block categories. In addition to supporting Dashicons, the idea is to introduce support for external components, functions and svg icons, in order to allow for third party plugins to be flexible with the icons for the categories they register. 

Since `<BlockIcon />` already supported all those kinds of methods to define an icon, we're abstracting the functionality to a separate component, and reusing it here and in `<BlockIcon />`.

## How has this been tested?
* Checkout this PR and build Gutenberg.
* Open the inserter in a post.
* Look at the "Common" category.
* Verify you can see the Dashicon to the right of the "Common" category name.
* Save at least one reusable block.
* Verify the reusable blocks category icon is displayed to the right of the category name.
* Verify that block categories without icons are unchanged.
* Verify existing block icons in the inserter are unchanged.
* Verify unit tests pass:
  * `BlockIcon` tests - `npm run test-unit packages/editor/src/components/block-icon/`
  * `Icon` tests - `npm run test-unit packages/components/src/icon/`
  * All unit tests - `npm run test-unit`
  * All tests - `npm test`

## Screenshots

Inserter with an icon on the "Common" block category:
~https://cldup.com/Eupr3BAnyc.png~
![](https://cldup.com/0MUzWwCDsT.png)

"Reusable" category in the inserter with one or more existing reusable blocks:
![](https://cldup.com/OlEso7ZItJ.png)

Inserter with a custom SVG icon on a custom category:
~https://cldup.com/etFPNTYIK3.png~
![](https://cldup.com/7HT6UgJyhG.png)

(see https://github.com/Automattic/wp-calypso/pull/27878 for testing it)

## Types of changes
* Introduced a new `<RawIcon />` component and exposed it.
* Added tests and documentation for `<RawIcon />`.
* Updated `<BlockIcon />` to use `<RawIcon />`.
* Updated the `PanelBody` to use `<RawIcon />` for block categories.
* Now passing category icon from inserter menu to `<PanelBody />`.
* Now registering the default categories with `null` icons (no icons).
* Now displaying the panel body icons on the right side of the panel body title.
* Moved and updated some tests.
* TO BE DELETED BEFORE MERGE - added a dashicon to the "Common" block category (https://github.com/WordPress/gutenberg/commit/204582b043c82ade5a3b24a2d9eb8130822dbf20).
* Renamed `<RawIcon />` to `<Icon />`.
* Made rendering of `Dashicon` default to a size of 20x20, while keeping 24x24 for the rest of the icon types.

## Checklist:

## Notes

* Since we've added a dashicon to the "Common" block category (https://github.com/WordPress/gutenberg/commit/204582b043c82ade5a3b24a2d9eb8130822dbf20) this needs to be deleted before merging.
